C++-accessible Placements via pybind11

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163667
* #162990
* #164414
* __->__ #163030

This makes Placement data representation available in C++ via pybind11.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @ezyang @chenyang78

swolchok committed:
C++-accessible Placements via pybind11

This makes Placement data representation available in C++ via pybind11.

[ghstack-poisoned]

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2886,6 +2888,119 @@ static void initTensorImplConversion(PyObject* module) {
 } // namespace torch::autograd
 
+// XXX: Move this somewhere
+static const char placement_class_docstring[] =
```
Should also be constexpr

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2886,6 +2888,119 @@ static void initTensorImplConversion(PyObject* module) {
+
+class Replicate : public Placement {
+ public:
+  ~Replicate() {}
```
Should be equals default, right?

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2886,6 +2888,119 @@ static void initTensorImplConversion(PyObject* module) {
+  explicit Partial(std::string reduce_op_) : reduce_op(std::move(reduce_op_)) {}
+
+  ~Partial() {}
+
```
Likewise here

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2886,6 +2888,119 @@ static void initTensorImplConversion(PyObject* module) {
+  }
+};
+
+Placement::~Placement() {}
```
Shouldn't this be equals default too? Why is it out of line?

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

[ghstack-poisoned]

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

[ghstack-poisoned]

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
need to pick up fix in previous pr on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
reorder #162990 after #163030 given current custom op uncertainty on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix lint (except weird linting of the cmath header),  python test/distributed/tensor/test_dtensor.py DTensorMeshTest.test_vmap_embedding, and python test/distributed/tensor/test_dtensor_compile.py TestDTensorCompile.test_dtensor_dynamic on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
lintrunner keeps putting a newline after import torch on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix a random Python Placement subclass on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix NormPartial again on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix NormPartial again, again on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix _StridedShard not inheriting from Shard on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix lintrunner mypy on 162990 on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ezyang commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -20,29 +21,11 @@
 __all__ = ["Placement", "Shard", "Replicate", "Partial"]
```
BTW there's another placement MaskPartial which should be moved over here

ezyang commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -142,7 +142,7 @@ def is_placement_type(value):
         from torch.distributed.tensor.placement_types import Placement
 
-        return type(value) is type and issubclass(value, Placement)
+        return isinstance(value, type) and issubclass(value, Placement)
```
lol

swolchok committed:
pickle to dicts for compatibility on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

malfet commented on torch/csrc/distributed/python_placement.cpp:
```diff
@@ -0,0 +1,98 @@
+namespace {
+// NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
+const char placement_class_docstring[] =
+    "The base class for the Placement type, where it describes how a DTensor is placed onto the\n"
+    "``DeviceMesh``. ``Placement`` and ``DeviceMesh`` together could describe the DTensor Layout.\n"
+    "It is the base class of the three main DTensor Placement types: ``Shard``, ``Replicate``,\n"
+    "and ``Partial``.\n"
+    "\n"
+    "This class is not meant to be used directly, mainly served as a typing stub.\n";
```
Multi-line string linterals are supported since C++11 
```suggestion
    R""""The base class for the Placement type, where it describes how a DTensor is placed onto the
   ``DeviceMesh``. ``Placement`` and ``DeviceMesh`` together could describe the DTensor Layout.
   It is the base class of the three main DTensor Placement types: ``Shard``, ``Replicate``,
    and ``Partial``.
    \n
    "This class is not meant to be used directly, mainly served as a typing stub.
""";
```

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
rebase on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok commented on torch/csrc/distributed/python_placement.h:
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <Python.h>
```
done -- we have a header that takes care of this

swolchok committed:
fix Replicate pickle implementation returning unusable object for some reason on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -r

swolchok committed:
rebase with conflicts on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "C++-accessible Placements via pybind11"


This makes Placement data representation available in C++ via pybind11.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok commented:
@pytorchbot revert -m "doesn't pass pyre" -c nosignal
